cdap-ui/app/cdap/components/DAG/DAGRenderer.tsx

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+*/
+
+import * as React from 'react';
+import 'jsplumb';
+import { INode, IConnection } from 'components/DAG/DAGProvider';
+import { DefaultNode } from 'components/DAG/Nodes/Default';
+import { List } from 'immutable';
+import uuidV4 from 'uuid/v4';
+
+const DAG_CONTAINER_ID = uuidV4();
+
+export interface IEndPointArgs {
+  element: HTMLElement | null;
+  params?: EndpointParams;
+  referenceParams?: EndpointParams;
+}
+
+export interface IInitNodeProps {
+  nodeId: string;
+  endPointParams?: IEndPointArgs[];
+  makeSourceParams?: any;
+  makeTargetParams?: any;
+}
+
+export interface IRegisterTypesProps {
+  connections: {
+    [anyProp: string]: any;
+  };
+  endpoints: {
+    [anyProp: string]: any;
+  };
+}
+
+export interface INodeComponentProps extends INode {
+  initNode?: (initConfig: IInitNodeProps) => void;
+  onDelete?: (nodeId: string) => void;
+}
+
+interface IDAGRendererProps {
+  nodes: List<INode>;
+  connections: List<IConnection>;
+  onConnection: (connection) => void;
+  onConnectionDetached: (connection) => void;
+  onDeleteNode: (nodeId: string) => void;
+  jsPlumbSettings: object;
+  registerTypes?: IRegisterTypesProps;
+}
+
+export class DAGRenderer extends React.Component<IDAGRendererProps, any> {
+  public state = {
+    isJsPlumbInstanceCreated: false,
+    jsPlumbInstance: jsPlumb.getInstance(this.props.jsPlumbSettings || {}),
+  };
+
+  public componentDidMount() {
+    jsPlumb.ready(() => {
+      const jsPlumbInstance = jsPlumb.getInstance({
+        Container: DAG_CONTAINER_ID,
+      });
+      jsPlumbInstance.setContainer(document.getElementById(DAG_CONTAINER_ID));
+      jsPlumbInstance.bind('connection', (connObj: IConnection, originalEvent: boolean) => {
+        if (!originalEvent) {
+          return;
+        }
+        const newConnObj = this.getNewConnectionObj(connObj);
+        this.props.onConnection(newConnObj);
+        this.state.jsPlumbInstance.repaintEverything();
+      });
+      jsPlumbInstance.bind('connectionDetached', (connObj: IConnection, originalEvent: boolean) => {
+        if (!originalEvent) {
+          return;
+        }
+        const newConnObj = this.getNewConnectionObj(connObj);
+        this.props.onConnectionDetached(newConnObj);
+        this.state.jsPlumbInstance.repaintEverything();
```

Comment:
do we need to repaint when endpoints get connected and disconnected?

Reply:
Fortunately no. Added to test something and missed that one there. Good catch.